hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java

```diff
@@ -0,0 +1,446 @@
+      String prev = snapshotChainPath
+          .get(snapshotPath)
+          .get(snapshotID)
+          .getPreviousSnapshotID();
+
+      if (prev != null &&
+          !snapshotChainPath
+              .get(snapshotPath)
+              .containsKey(prev)) {
+        throw new IOException("Snapshot chain corruption: snapshot node to "
+                + "be deleted has prev node element not found in snapshot "
+                + "chain: Snapshot path " + snapshotPath + ", SnapshotID "
+                + prev);
+      }
+      if (next != null && !snapshotChainPath
+          .get(snapshotPath)
+          .containsKey(next)) {
+        throw new IOException("Snapshot chain corruption: snapshot node to "
+                + "be deleted has next node element not found in snapshot "
+                + "chain:  Snapshot path " + snapshotPath + ", SnapshotID "
+                + next);
+      }
+      snapshotChainPath
+          .get(snapshotPath)
+          .remove(snapshotID);
+      if (next != null) {
+        snapshotChainPath
+            .get(snapshotPath)
+            .get(next)
+            .setPreviousSnapshotID(prev);
+      }
+      if (prev != null) {
+        snapshotChainPath
+            .get(snapshotPath)
+            .get(prev)
+            .setNextSnapshotID(next);
+      }
+      // remove from latest list if necessary
+      if (latestPathSnapshotID.get(snapshotPath).equals(snapshotID)) {
+        latestPathSnapshotID.remove(snapshotPath);
+        if (prev != null) {
+          latestPathSnapshotID.put(snapshotPath, prev);
+        }
+      }
+
+    } else {
+      // snapshotID not found in snapshot chain, log warning and return
+      LOG.warn("Snapshot chain: snapshotID not found: Snapshot path {}," +
+              " SnapshotID {}",
+          snapshotPath, snapshotID);
+    }
+
+    return status;
+  }
+
+  private void loadFromSnapshotInfoTable(OMMetadataManager metadataManager)
+          throws IOException {
+    // read from snapshotInfo table to populate
+    // snapshot chains - both global and local path
+    TableIterator<String, ? extends Table.KeyValue<String, SnapshotInfo>>
+            keyIter = metadataManager.getSnapshotInfoTable().iterator();
+    Map<Long, SnapshotInfo> snaps = new TreeMap<>();
+    Table.KeyValue< String, SnapshotInfo > kv;
+    snapshotChainGlobal.clear();
+    snapshotChainPath.clear();
+
+    while (keyIter.hasNext()) {
+      kv = keyIter.next();
+      snaps.put(kv.getValue().getCreationTime(), kv.getValue());
+    }
+    for (SnapshotInfo sinfo : snaps.values()) {
+      addSnapshot(sinfo);
+    }
+  }
+
+  public void addSnapshot(SnapshotInfo sinfo) throws IOException {
+    addSnapshotGlobal(sinfo.getSnapshotID(),
+        sinfo.getGlobalPreviousSnapshotID());
+    addSnapshotPath(sinfo.getSnapshotPath(),
+        sinfo.getSnapshotID(),
+        sinfo.getPathPreviousSnapshotID());
+  }
+
+  public boolean deleteSnapshot(SnapshotInfo sinfo) throws IOException {
```

Comment:
does this ever return false?

Reply:
Will fix.  Need to set status as well as log warning in deletes, line 167 and line 234.